controller/internal/enforcer/envoyauthorizer/envoyproxy/auth_server.go

```diff
@@ -0,0 +1,466 @@
+	var sourceIP, destIP string
+	var source, dest *ext_auth.AttributeContext_Peer
+	var httpReq *ext_auth.AttributeContext_HttpRequest
+	var destPort, srcPort int
+	var urlStr, method string
+	attrs := checkRequest.GetAttributes()
+	if attrs != nil {
+		source = attrs.GetSource()
+		dest = attrs.GetDestination()
+
+		if source != nil {
+			if addr := source.GetAddress(); addr != nil {
+				if sockAddr := addr.GetSocketAddress(); sockAddr != nil {
+					sourceIP = sockAddr.GetAddress()
+					srcPort = int(sockAddr.GetPortValue())
+				}
+			}
+		}
+		if dest != nil {
+			if destAddr := dest.GetAddress(); destAddr != nil {
+				if destSockAddr := destAddr.GetSocketAddress(); destSockAddr != nil {
+					destIP = destSockAddr.GetAddress()
+					destPort = int(destSockAddr.GetPortValue())
+				}
+			}
+		}
+
+		if request := attrs.GetRequest(); request != nil {
+			httpReq = request.GetHttp()
+			urlStr = httpReq.GetPath()
+			method = httpReq.GetMethod()
+		}
+	}
+	// Create the new target URL based on the path parameter that we have from envoy.
+	URL, err := url.ParseRequestURI(urlStr)
+	if err != nil {
+		zap.L().Debug("ext_authz egress: Cannot parse the URI")
+		return nil, nil
+	}
+
+	authRequest := &apiauth.Request{
+		OriginalDestination: &net.TCPAddr{IP: net.ParseIP(destIP), Port: destPort},
+		SourceAddress:       &net.TCPAddr{IP: net.ParseIP(sourceIP), Port: srcPort},
+		URL:                 URL,
+		Method:              method,
+		RequestURI:          "",
+	}
+	r := new(http.Request)
+	r.RemoteAddr = sourceIP
+	resp, err := s.auth.ApplicationRequest(authRequest)
+	if err != nil {
+		if resp.PUContext != nil {
+			state := flowstats.NewAppConnectionState(s.puID, r, authRequest, resp)
+			state.Stats.Action = resp.Action
+			state.Stats.PolicyID = resp.NetworkPolicyID
+			s.collector.CollectFlowEvent(state.Stats)
+		}
+		//http.Error(w, err.Error(), err.(*apiauth.AuthError).Status())
+		zap.L().Debug("ext_authz egress: Access *NOT* authorized by network policy", zap.String("puID", s.puID))
+		//flow.DropReason = "access not authorized by network policy"
+		return createDeniedCheckResponse(rpc.PERMISSION_DENIED, envoy_type.StatusCode_Forbidden, "Access not authorized by network policy"), err
+	}
+
+	state := flowstats.NewAppConnectionState(s.puID, r, authRequest, resp)
+	if resp.External {
+		defer s.collector.CollectFlowEvent(state.Stats)
+	}
+	if resp.Action.Rejected() {
+		zap.L().Debug("ext_authz egress: Access *NOT* authorized by network policy", zap.String("puID", s.puID))
+		//flow.DropReason = "access not authorized by network policy"
+		return createDeniedCheckResponse(rpc.PERMISSION_DENIED, envoy_type.StatusCode_Forbidden, "Access not authorized by network policy"), nil
+	}
+	// now create the response and inject our identity
+	zap.L().Debug("ext_authz egress: injecting header", zap.String("puID", s.puID))
+	// build our identity token
+	var transmittedKey []byte
+	if s.secrets != nil {
+		transmittedKey = s.secrets.TransmittedKey()
```

Comment:
why would secrets be nil?

anyway, there is a potential problem with accessing secrets here ... that's exactly why I added the `Secrets()` function on the enforcer. All those gRPC servers are running in their own threads/goroutines. So accessing the secret must be done with care.